c2rust-analyze/src/rewrite/ty.rs

```diff
@@ -0,0 +1,370 @@
+            if hir_mt.mutbl == convert_mutability(tm.mutbl) {
+                Some(vec![hir_mt.ty])
+            } else {
+                None
+            }
+        }
+        (&TyKind::Ref(_, _, mutbl), &hir::TyKind::Rptr(_, ref hir_mt)) => {
+            if hir_mt.mutbl == convert_mutability(mutbl) {
+                Some(vec![hir_mt.ty])
+            } else {
+                None
+            }
+        }
+        (&TyKind::Tuple(tys), &hir::TyKind::Tup(hir_tys)) => {
+            if tys.len() == hir_tys.len() {
+                Some(hir_tys.iter().collect())
+            } else {
+                None
+            }
+        }
+
+        _ => None,
+    }
+}
+
+fn convert_int_ty(x: ty::IntTy) -> ast::IntTy {
+    match x {
+        ty::IntTy::Isize => ast::IntTy::Isize,
+        ty::IntTy::I8 => ast::IntTy::I8,
+        ty::IntTy::I16 => ast::IntTy::I16,
+        ty::IntTy::I32 => ast::IntTy::I32,
+        ty::IntTy::I64 => ast::IntTy::I64,
+        ty::IntTy::I128 => ast::IntTy::I128,
+    }
+}
+
+fn convert_uint_ty(x: ty::UintTy) -> ast::UintTy {
+    match x {
+        ty::UintTy::Usize => ast::UintTy::Usize,
+        ty::UintTy::U8 => ast::UintTy::U8,
+        ty::UintTy::U16 => ast::UintTy::U16,
+        ty::UintTy::U32 => ast::UintTy::U32,
+        ty::UintTy::U64 => ast::UintTy::U64,
+        ty::UintTy::U128 => ast::UintTy::U128,
+    }
+}
+
+fn convert_float_ty(x: ty::FloatTy) -> ast::FloatTy {
+    match x {
+        ty::FloatTy::F32 => ast::FloatTy::F32,
+        ty::FloatTy::F64 => ast::FloatTy::F64,
+    }
+}
+
+fn convert_mutability(x: mir::Mutability) -> hir::Mutability {
+    match x {
+        mir::Mutability::Mut => hir::Mutability::Mut,
+        mir::Mutability::Not => hir::Mutability::Not,
+    }
+}
+
+fn mk_cell<'tcx>(tcx: TyCtxt<'tcx>, ty: Ty<'tcx>) -> Ty<'tcx> {
+    let core_crate = tcx
+        .crates(())
+        .iter()
+        .cloned()
+        .find(|&krate| tcx.crate_name(krate).as_str() == "core")
+        .expect("failed to find crate `core`");
+
+    let cell_mod_child = tcx
+        .module_children(core_crate.as_def_id())
+        .iter()
+        .find(|child| child.ident.as_str() == "cell")
+        .expect("failed to find module `core::cell`");
+    let cell_mod = match cell_mod_child.res {
+        Res::Def(DefKind::Mod, did) => did,
+        ref r => panic!("unexpected resolution {:?} for `core::cell`", r),
+    };
+
+    let cell_struct_child = tcx
+        .module_children(cell_mod)
+        .iter()
+        .find(|child| child.ident.as_str() == "Cell")
+        .expect("failed to find struct `core::cell::Cell`");
+    let cell_struct = match cell_struct_child.res {
+        Res::Def(DefKind::Struct, did) => did,
+        ref r => panic!("unexpected resolution {:?} for `core::cell::Cell`", r),
+    };
+
+    let cell_adt = tcx.adt_def(cell_struct);
+    let substs = tcx.mk_substs([GenericArg::from(ty)].into_iter());
+    tcx.mk_adt(cell_adt, substs)
+}
```

Comment:
I think the proper (and much easier and faster) way to do this is using diagnostic items.  We can call `tcx.get_diagnostic_item(Symbol::intern(...))` and get the `DefId` directly.  Only `Cell` doesn't seem to have one, which I'm surprised, since I would've thought it'd be used in some diagnostic.  `RefCell` does, though: https://github.com/rust-lang/rust/blob/068161ea483b1a80a959476cb3e31e6619a72737/library/core/src/cell.rs#L628-L630.  So you can at least skip most of this function to start at `RefCell`.

Reply:
Oh, that was added on 2022-09-27, and we're on 2022-08-08.  We can similarly use `tcx.lang_items().unsafe_cell_type()`.

Reply:
Seems weird to depend on which types happen to be mentioned in some diagnostic suggestion.  I think we should just define a helper function for this lookup (once we're doing this for more than one type), and maybe add some caching if it turns out to be a performance problem.

Anyway, since `Cell` isn't a diagnostic item at the moment, I don't think there's anything to change here for now.